### PR TITLE
feat(ci)!: implement reusable workflows for image building

### LINKED ISSUE #4106: ci: refactor most pipelines to make use of reusable workflows
## Motivation

We're not following the DRY principle with the pipelines, as most workflows have at least a 20% of repeated steps, and we're moving into a stage were we need to have more workflow segregation, for example moving all the disk rebuild steps to their own workflow, but by doing this we would also create even more repetition and code harder to maintain.

### Specifications

https://docs.github.com/cn/actions/using-workflows/reusing-workflows

### Designs
- Create a workflow for building docker images (`docker-build.yml`)
- Make other workflows that require a docker image to run tests or for deployment depend on `docker-build.yml`
- Move disk rebuild steps/jobs to their own workflow, allowing to run this workflow without triggering other tests
- Merge the Full sync workflow with the Integration Tests as the paths used to run this workflows is basically the same

## Related Work
[refactor(actions): rename workflow files](https://github.com/ZcashFoundation/zebra/pull/3941#) #3941
[Skip duplicate GitHub actions](https://github.com/ZcashFoundation/zebra/issues/1786#) #1786

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation

We're not following the DRY principle with the pipelines, as most workflows have at least a 20% of repeated steps, and we're moving into a stage were we need to have more workflow segregation, for example moving all the disk rebuild steps to their own workflow, but by doing this we would also create even more repetition and code harder to maintain.

Fixes #4106

### Specifications

https://docs.github.com/cn/actions/using-workflows/reusing-workflows

### Designs

- Use a single builder for all workflows using Dockerfile

## Solution

- Create a single `docker-zebra-build.yml` workflow which is used to build docker images whichever the target is (production, test, zcash-params)
- Make other build steps to depend on this `build` workflow

## Review

@dconnolly can review this

### Reviewer Checklist

  - [ ] Verify the `workflow_call` makes sense
  - [ ] Confirm the `inputs` being sent as defaults from the `build` step available in `test.yml`, `test-full-sync.yml`, `cd.yml` and `zcash-params.yml`
  - [ ] Tests for Errors

## Follow Up Work

- lightwalletd keeps the old approach as it depends on an external repository checkout to build correctly.

### CODE REVIEW

File: .github/workflows/cd.yml [hunk 1]
```diff
@@ -16,113 +16,26 @@ on:
...
-            SHORT_SHA=${{ env.GITHUB_SHA_SHORT }}
-            RUST_BACKTRACE=1
-            ZEBRA_SKIP_IPV6_TESTS="1"
-            CHECKPOINT_SYNC=${{ github.event.inputs.checkpoint_sync || true }}
-            SENTRY_DSN=${{ secrets.SENTRY_ENDPOINT }}
-          push: true
-          cache-from: type=registry,ref=${{ env.GAR_BASE }}/${{ env.GITHUB_REF_SLUG_URL }}:buildcache
-          cache-to: type=registry,ref=${{ env.GAR_BASE }}/${{ env.GITHUB_REF_SLUG_URL }}:buildcache,mode=max
+    # TODO: before merging change uses to ZcashFoundation/zebra/.github/workflows/docker-zebra-build.yml@main
+    uses: ./.github/workflows/docker-zebra-build.yml
```
Comment (teor2345) [inline .github/workflows/cd.yml]: Question: What if we want to change this workflow and its dependent workflows in a PR? Would we be better to always use the version in the current commit?

File: .github/workflows/cd.yml [hunk 2]
```diff
@@ -16,113 +16,26 @@ on:
...
+    with:
+      # TODO validate if CARGO_INCREMENTAL should be added at build time
+      dockerfile_path: ./docker/Dockerfile
+      dockerfile_target: runtime
+      image_name: main
```
Comment (teor2345) [inline .github/workflows/cd.yml]: Question: Did this used to be the branch name? What if we're using workflow_dispatch, and the branch we're running this workflow on isn't the `main` branch?

File: .github/workflows/cd.yml [hunk 3]
```diff
@@ -16,113 +16,26 @@ on:
...
-            ZEBRA_SKIP_IPV6_TESTS="1"
-            CHECKPOINT_SYNC=${{ github.event.inputs.checkpoint_sync || true }}
-            SENTRY_DSN=${{ secrets.SENTRY_ENDPOINT }}
-          push: true
-          cache-from: type=registry,ref=${{ env.GAR_BASE }}/${{ env.GITHUB_REF_SLUG_URL }}:buildcache
-          cache-to: type=registry,ref=${{ env.GAR_BASE }}/${{ env.GITHUB_REF_SLUG_URL }}:buildcache,mode=max
+    # TODO: before merging change uses to ZcashFoundation/zebra/.github/workflows/docker-zebra-build.yml@main
+    uses: ./.github/workflows/docker-zebra-build.yml
+    with:
+      # TODO validate if CARGO_INCREMENTAL should be added at build time
```
Comment (teor2345) [inline .github/workflows/cd.yml]: `CARGO_INCREMENTAL` can decrease build time when running from a cache. But it also reduces compiler optimisations.

Right now, I'd leave it off all our builds, because runtime speed is a bigger problem than compilation speed.

File: .github/workflows/docker-zebra-build.yml [hunk 4]
```diff
@@ -0,0 +1,125 @@
+name: Full sync test
+
+on:
+  workflow_call:
+    inputs:
+      network:
```
Comment (teor2345) [inline .github/workflows/docker-zebra-build.yml]: Possible Blocker: there are a bunch of lints here that might need fixing.

File: .github/workflows/test-full-sync.yml [hunk 5]
```diff
@@ -45,111 +45,34 @@ on:
       - '.github/workflows/test-full-sync.yml'
```
Comment (teor2345) [inline .github/workflows/test-full-sync.yml:45]: Blocker: please add all the dependent workflow files to all instances of the build filter paths.

To make sure we don't miss any, it might be best to use a search and replace.
I quite like `fastmod`, but I'm not sure if it has a multi-line mode.

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -51,107 +51,30 @@ on:
       - '.github/workflows/test.yml'
```
Comment (teor2345) [inline .github/workflows/test.yml:51]: Blocker: please add all the dependent workflow files to all instances of the build filter paths.

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -51,107 +51,30 @@ on:
...
-            RUST_LOG=debug
-            SENTRY_DSN=${{ secrets.SENTRY_ENDPOINT }}
-          push: true
-          cache-from: type=registry,ref=${{ env.GAR_BASE }}/${{ env.IMAGE_NAME }}:${{ env.GITHUB_REF_SLUG_URL }}-buildcache
-          cache-to: type=registry,ref=${{ env.GAR_BASE }}/${{ env.IMAGE_NAME }}:${{ env.GITHUB_REF_SLUG_URL }}-buildcache,mode=max
+    # TODO: before merging change uses to ZcashFoundation/zebra/.github/workflows/docker-zebra-build.yml@main
+    uses: ./.github/workflows/docker-zebra-build.yml
+    with:
+      # TODO validate if CARGO_INCREMENTAL should be added at build time
+      dockerfile_path: ./docker/Dockerfile
```
Comment (teor2345) [inline .github/workflows/test.yml:67]: Question: These arguments are exactly the same as in `test-full-sync.yml`. So I'm trying to understand how they work together, and if they can conflict.

When a branch is pushed, does it always create just one image? Or do we build two images with the same name, and the last one wins?

File: .github/workflows/zcash-params.yml [hunk 8]
```diff
@@ -14,98 +14,18 @@ on:
...
+      dockerfile_path: ./docker/zcash-params/Dockerfile
+      dockerfile_target: builder
+      image_name: zcash-params
+      network: Mainnet
+      checkpoint_sync: true
+      rust_backtrace: full
+      rust_lib_backtrace: full
+      colorbt_show_hidden: '1'
+      zebra_skip_network_tests: '1'
+      rust_log: debug
```
Comment (teor2345) [inline .github/workflows/zcash-params.yml]: Question: Some of these parameters are unused. Can we document that, to avoid confusion?

```yml
      # only used in tests, this is a build
      network: Mainnet
      checkpoint_sync: true
      zebra_skip_network_tests: '1'
```

File: .github/workflows/zcash-params.yml [hunk 9]
```diff
@@ -14,98 +14,18 @@ on:
       - 'docker/zcash-params/Dockerfile'
       - '.github/workflows/zcash-params.yml'
```
Comment (teor2345) [inline .github/workflows/zcash-params.yml:15]: Blocker: please add all the dependent workflow files to all instances of the build filter paths.

File: zebra-state/src/request.rs [hunk 10]
```diff
@@ -6,6 +6,7 @@ use std::{
     sync::Arc,
 };
 
+
```
Comment (teor2345) [inline zebra-state/src/request.rs]: Blocker: Please revert this change before merging, it will cause `rustfmt` to fail.

As a longer-term fix, can you open a ticket to add the relevant workflow and docker files to the workflow path filters?

File: zebra-test/Cargo.toml [hunk 11]
```diff
@@ -31,3 +31,4 @@ tracing = "0.1.31"
 
 [dev-dependencies]
 tempfile = "3.3.0"
+
```
Comment (teor2345) [inline zebra-test/Cargo.toml]: Minor: Can you revert this change before merging, it might cause merge conflicts?

As a longer-term fix, can you open a ticket to add the relevant workflow and docker files to the workflow path filters?

**Review by teor2345 — COMMENTED** [review]
Looks like a good start, but I have some questions about how it works.

File: .github/workflows/cd.yml (hunk 2 above)
Comment (teor2345) [reply .github/workflows/cd.yml]: Maybe we want something like:
```suggestion
      image_name: zebrad-deploy
```

File: .github/workflows/cd.yml (hunk 1 above)
Comment (gustavovalverde) [reply .github/workflows/cd.yml]: This is more for stability purposes, but keeping it as-is right now is more convenient. Maybe we can leave the TODO and change the text a bit for reconsideration instead of not merging it like this.

File: .github/workflows/cd.yml (hunk 2 above)
Comment (gustavovalverde) [reply .github/workflows/cd.yml]: Great question, as I even considered removing the  **workflow_dispatch** from the Continuous deployment workflow. I don't think we wan't to allow deployments to _"Mainnet"_ from a random branch. 

At worst, because of a deployment failure, we just want to rerun the failed job.

Comment (gustavovalverde) [reply .github/workflows/cd.yml]: The image name has always been `main` (as it takes the name from the branch), but we can surely change it to `production`,  plain`zebrad` or `zebrad-deploy` as you suggested.

@dconnolly can help us here for taking the decision.

File: .github/workflows/docker-zebra-build.yml (hunk 4 above)
Comment (gustavovalverde) [reply .github/workflows/docker-zebra-build.yml]: A `workflow_call` does not need a description for each input, we can add it for documentation purposes, but the linter is not considering that this is not a `workflow_dispatch` but a `workflow_call`

Adding the description is painless, but I'd like some help for some of those if we agree on adding it.

File: .github/workflows/cd.yml (hunk 1 above)
Comment (gustavovalverde) [reply .github/workflows/cd.yml]: Decided to remove the TODO as it might be confusing anyways.

File: .github/workflows/zcash-params.yml (hunk 8 above)
Comment (gustavovalverde) [reply .github/workflows/zcash-params.yml]: I removed the unused ones

File: .github/workflows/cd.yml (hunk 2 above)
Comment (dconnolly) [reply .github/workflows/cd.yml]: I think this has inherited the branch name, so that things had the branch they built off of (which you can also get for workflow_dispatch)

File: .github/workflows/test.yml (hunk 7 above)
Comment (gustavovalverde) [reply .github/workflows/test.yml:67]: This is a great question, and I realized the same and plan to refactor both of this workflows so we don't build the image twice, it won't conflict (this is happening now when both workflows run at the same time) but it uses extra CI time and also creates more artifacts in GCP.

It has the same behavior as now, but we should simplify this. I'm not doing it in this PR as it would make things more complex to understand

**Comment by teor2345** [comment]
@mergifyio update

**Comment by teor2345** [comment]
I am updating this branch to remove the clean action, which causes other actions to fail.

File: .github/workflows/cd.yml (hunk 2 above)
Comment (gustavovalverde) [reply .github/workflows/cd.yml]: Made `zebrad` the image name

File: .github/workflows/test.yml (hunk 7 above)
Comment (teor2345) [reply .github/workflows/test.yml:67]: Is there a ticket for this follow-up?

File: .github/workflows/docker-image-build.yml [hunk 12]
```diff
@@ -0,0 +1,125 @@
+name: Full sync test
```
Comment (teor2345) [inline .github/workflows/docker-image-build.yml]:
```suggestion
name: Build docker image
```

File: .github/workflows/test-full-sync.yml [hunk 13]
```diff
@@ -43,113 +43,35 @@ on:
...
+      dockerfile_path: ./docker/Dockerfile
+      dockerfile_target: tester
+      image_name: zebrad-test
+      network: Mainnet
+      checkpoint_sync: true
+      rust_backtrace: full
+      rust_lib_backtrace: full
+      colorbt_show_hidden: '1'
+      zebra_skip_ipv6_tests: '1'
+      rust_log: debug
```
Comment (teor2345) [inline .github/workflows/test-full-sync.yml]: Performance: debug logging slows down Zebra a lot. Let's only turn it on if we really need it. I'll open a separate PR for this change, because #4155 is urgent.

```suggestion
      rust_log: info
```

File: .github/workflows/test.yml [hunk 14]
```diff
@@ -49,109 +49,31 @@ on:
...
+      dockerfile_target: tester
+      image_name: zebrad-test
+      network: Mainnet
+      checkpoint_sync: true
+      rust_backtrace: full
+      rust_lib_backtrace: full
+      colorbt_show_hidden: '1'
+      zebra_skip_ipv6_tests: '1'
+      rust_log: debug
```
Comment (teor2345) [inline .github/workflows/test.yml]:
```suggestion
      rust_log: info
```

File: .github/workflows/zcash-params.yml [hunk 15]
```diff
@@ -13,99 +13,16 @@ on:
...
-          cache-to: type=registry,ref=${{ env.GAR_BASE }}/${{ env.IMAGE_NAME }}:buildcache,mode=max
+    uses: ./.github/workflows/docker-image-build.yml
+    with:
+      dockerfile_path: ./docker/zcash-params/Dockerfile
+      dockerfile_target: builder
+      image_name: zcash-params
+      rust_backtrace: full
+      rust_lib_backtrace: full
+      colorbt_show_hidden: '1'
+      rust_log: debug
```
Comment (teor2345) [inline .github/workflows/zcash-params.yml]:
```suggestion
      rust_log: info
```

**Review by teor2345 — COMMENTED** [review]
Looks great, just some final tweaks

File: .github/workflows/test.yml (hunk 7 above)
Comment (gustavovalverde) [reply .github/workflows/test.yml:67]: I'm adding a task for myself to create this ticket tomorrow morning.

**Comment by gustavovalverde** [comment]
@teor2345 suggestions applied.

**Comment by gustavovalverde** [comment]
I'd like to use this approach on the `lightwalletd` tests. Can we merge this one first and update [lwd-sync-test](https://github.com/ZcashFoundation/zebra/tree/lwd-sync-test) with it?

**Review by teor2345 — APPROVED** [review]
> I'd like to use this approach on the `lightwalletd` tests. Can we merge this one first and update [lwd-sync-test](https://github.com/ZcashFoundation/zebra/tree/lwd-sync-test) with it?

There's nothing urgent pending, so let's give it a shot.

(Worst case scenario, if this PR makes `main` fail, we can revert it and try again.)

**Comment by teor2345** [comment]
@mergifyio refresh
